### ISSUE TITLE
Exporting using s3 table function might create too many threads

### ISSUE THREAD

**Issue by Algunenano** [ClickHouse/ClickHouse#51533]
Create a table to export it to s3 using partitions:

```
Mordor :) create table n (a Int64, b Int64) ENGINE=MergeTree() order by a;

CREATE TABLE n
(
    `a` Int64,
    `b` Int64
)
ENGINE = MergeTree
ORDER BY a

Query id: 9307f2ba-a356-481e-9055-8ccd4a30a29c

Ok.

0 rows in set. Elapsed: 0.001 sec. 

Mordor :) insert into n Select number, number from numbers(10000)

INSERT INTO n SELECT
    number,
    number
FROM numbers(10000)

Query id: 7c056482-30fc-4b27-a90f-a21ae6cede7e

Ok.

0 rows in set. Elapsed: 0.002 sec. Processed 10.00 thousand rows, 80.00 KB (6.00 million rows/s., 47.99 MB/s.)
```

Metrics before:
```
SELECT *
FROM system.asynchronous_metrics
WHERE (metric LIKE 'jemalloc%') OR (metric ILIKE '%thread%')
ORDER BY metric ASC

Query id: 0d0a854a-9013-4793-ae71-c9efb0a3d609

┌─metric───────────────────────────────────┬─────value─┬─description────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ HTTPThreads                              │         0 │ Number of threads in the server of the HTTP interface (without TLS).                                                                       │
│ MySQLThreads                             │         0 │ Number of threads in the server of the MySQL compatibility protocol.                                                                       │
│ OSThreadsRunnable                        │         4 │ The total number of 'runnable' threads, as the OS kernel scheduler seeing it.                                                              │
│ OSThreadsTotal                           │      3233 │ The total number of threads, as the OS kernel scheduler seeing it.                                                                         │
│ TCPThreads                               │         1 │ Number of threads in the server of the TCP protocol (without TLS).                                                                         │
│ jemalloc.active                          │ 118517760 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.allocated                       │ 113522904 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.arenas.all.dirty_purged         │         0 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.arenas.all.muzzy_purged         │         0 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.arenas.all.pactive              │     28935 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.arenas.all.pdirty               │     18114 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.arenas.all.pmuzzy               │         0 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.background_thread.num_runs      │         0 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.background_thread.num_threads   │         0 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.background_thread.run_intervals │         0 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.epoch                           │        17 │ An internal incremental update number of the statistics of jemalloc (Jason Evans' memory allocator), used in all other `jemalloc` metrics. │
│ jemalloc.mapped                          │ 276598784 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.metadata                        │  20653712 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.metadata_thp                    │         0 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.resident                        │ 205963264 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.retained                        │  69955584 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
└──────────────────────────────────────────┴───────────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

Try a bad insert into s3:

```
INSERT INTO
            TABLE FUNCTION s3(
                'http://resolver:8083/root/data/test-upload_{_partition_id}.csv.gz',
                'minio', 'minio123',
                'CSV', auto, 'gzip'
            )
        PARTITION BY (a, b) Select * from n;

INSERT INTO FUNCTION s3('http://resolver:8083/root/data/test-upload_{_partition_id}.csv.gz', 'minio', 'minio123', 'CSV', auto, 'gzip') PARTITION BY (a, b) SELECT *
FROM n

Query id: 53867b93-9b7a-4fa7-987c-df4123b8ad37

↗ Progress: 10.00 thousand rows, 160.00 KB (5.49 thousand rows/s., 87.77 KB/s.)  99%
0 rows in set. Elapsed: 1.823 sec. Processed 10.00 thousand rows, 160.00 KB (5.49 thousand rows/s., 87.77 KB/s.)

Received exception from server (version 23.6.1):
Code: 439. DB::Exception: Received from localhost:9000. DB::Exception: Cannot schedule a task: cannot allocate thread (threads=0, jobs=0). (CANNOT_SCHEDULE_TASK)
```

Metrics after:
```
SELECT *
FROM system.asynchronous_metrics
WHERE (metric LIKE 'jemalloc%') OR (metric ILIKE '%thread%')
ORDER BY metric ASC

Query id: 623f1b5f-43b5-49c2-9821-ddf8497f252d

┌─metric───────────────────────────────────┬───────value─┬─description────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ HTTPThreads                              │           0 │ Number of threads in the server of the HTTP interface (without TLS).                                                                       │
│ MySQLThreads                             │           0 │ Number of threads in the server of the MySQL compatibility protocol.                                                                       │
│ OSThreadsRunnable                        │           2 │ The total number of 'runnable' threads, as the OS kernel scheduler seeing it.                                                              │
│ OSThreadsTotal                           │        4220 │ The total number of threads, as the OS kernel scheduler seeing it.                                                                         │
│ TCPThreads                               │           1 │ Number of threads in the server of the TCP protocol (without TLS).                                                                         │
│ jemalloc.active                          │  1529765888 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.allocated                       │  1405988368 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.arenas.all.dirty_purged         │      216494 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.arenas.all.muzzy_purged         │       42074 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.arenas.all.pactive              │      373478 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.arenas.all.pdirty               │     4328316 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.arenas.all.pmuzzy               │      171949 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.background_thread.num_runs      │           0 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.background_thread.num_threads   │           0 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.background_thread.run_intervals │           0 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.epoch                           │          35 │ An internal incremental update number of the statistics of jemalloc (Jason Evans' memory allocator), used in all other `jemalloc` metrics. │
│ jemalloc.mapped                          │ 20317270016 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.metadata                        │   282603632 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.metadata_thp                    │           0 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.resident                        │ 19505512448 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
│ jemalloc.retained                        │  2280067072 │ An internal metric of the low-level memory allocator (jemalloc). See https://jemalloc.net/jemalloc.3.html                                  │
└──────────────────────────────────────────┴─────────────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

21 rows in set. Elapsed: 0.001 sec.
```

The server now holds an extra ~1000 threads an 16 GB of reserved memory.

Calling the export function again does not raise more threads, but the memory usage does increase with each call and it doesn't appear to completely go down. In my case it seems to estabilize at around 23-24 GB of memory on an empty database (default no config).

Note that I don't have minio running, so the url is invalid.

A couple of opinions:
* 1000 threads, or whatever, to export to s3 seems too much, no matter how bad the partition was. There should be a reasonable upper limit.
* Keeping those thousand threads alive is expensive in terms of memory

**Comment by Beetelbrox** [comment]
We hit this issue, or at least something very similar, a couple of days ago when a customer tried to export a small-ish table into ~2K files with a too granular partition key. The query ended up failing with `[Error] Memory limit (for query) exceeded: would use 16.02 GiB (attempt to allocate chunk of 4314972 bytes), maximum: 16.00 GiB. (MEMORY_LIMIT_EXCEEDED)` and when checking the query log it reported an abnormally large amount of threads and `peak_threads_usage: 2153`. Setting `max_threads` for the query seemed to limit the number of threads used but still got the memory error (which may or may not be unrelated).

When they retried some of the queries failed with an `[Error] Cannot schedule a task: cannot allocate thread (threads=1, jobs=1). (CANNOT_SCHEDULE_TASK)` error as above.

I'd be interested on trying to fix this issue, any advice about where to start looking? Thanks!

**Comment by Beetelbrox** [comment]
I've found out that for some reason the large `peak_threads_usage` does NOT happen when exporting to Parquet:

Query log fragment:
```
SELECT
    query,
    formatReadableSize(read_bytes),
    formatReadableSize(written_bytes),
    formatReadableSize(memory_usage) AS used_memory,
    peak_threads_usage,
    ProfileEvents['S3PutObject'] AS files,
    Settings,
    ProfileEvents
FROM system.query_log
WHERE (event_time >= (now() - toIntervalHour(1))) AND (query_id = '60274550-eafe-4b3a-94cf-6f0536560dac') AND (type > 1)
FORMAT Vertical

Query id: 7332feeb-6c98-4989-a616-d82ad3f5cd30

Row 1:
──────
query:                             INSERT INTO FUNCTION s3('s3://test-sinks/test-threads/{_partition_id}.parquet', 'REDACTED', '[HIDDEN]') PARTITION BY number SELECT * FROM numbers(1000)
formatReadableSize(read_bytes):    7.81 KiB
formatReadableSize(written_bytes): 7.81 KiB
used_memory:                       1011.63 MiB
peak_threads_usage:                3
files:                             1000
Settings:                          {'max_threads':'16','use_uncompressed_cache':'0','distributed_foreground_insert':'1','move_all_conditions_to_prewhere':'0','enable_multiple_prewhere_read_steps':'0','move_primary_key_columns_to_end_of_prewhere':'0','compile_expressions':'0','compile_aggregate_expressions':'1','enable_memory_bound_merging_of_aggregation_results':'0','cluster_for_parallel_replicas':'gcp_wadus_4','insert_deduplicate':'1','http_wait_end_of_query':'1','http_write_exception_in_output_format':'0','http_response_buffer_size':'104857600','joined_subquery_requires_alias':'0','max_bytes_before_external_group_by':'1442450940','max_execution_time':'0','max_expanded_ast_elements':'50000','join_algorithm':'hash','max_memory_usage':'5000000000','memory_usage_overcommit_max_wait_microseconds':'50000','log_query_threads':'0','formatdatetime_f_prints_single_zero':'1','formatdatetime_parsedatetime_m_is_month_name':'0','max_partitions_per_insert_block':'100','enable_lightweight_delete':'0','optimize_monotonous_functions_in_order_by':'0','enable_global_with_statement':'0','optimize_rewrite_sum_if_to_count_if':'0','distributed_ddl_entry_format_version':'3','enable_filesystem_cache':'1','enable_filesystem_cache_on_write_operations':'1','enable_filesystem_cache_log':'1','allow_prefetched_read_pool_for_remote_filesystem':'1','filesystem_prefetch_max_memory_usage':'5368709120','insert_keeper_max_retries':'15','insert_keeper_retry_initial_backoff_ms':'100','insert_keeper_retry_max_backoff_ms':'2000','input_format_null_as_default':'0'}
ProfileEvents:                     {'Query':1,'InsertQuery':1,'QueriesWithSubqueries':2,'SelectQueriesWithSubqueries':2,'IOBufferAllocs':2000,'IOBufferAllocBytes':2097278000,'ArenaAllocChunks':1,'ArenaAllocBytes':4096,'FunctionExecute':2,'TableFunctionExecute':2,'NetworkReceiveElapsedMicroseconds':5211,'NetworkSendElapsedMicroseconds':321047,'NetworkSendBytes':1447974,'InsertedRows':1000,'InsertedBytes':8000,'SelectedRows':1000,'SelectedBytes':8000,'ContextLock':1051,'RWLockAcquiredReadLocks':4,'RealTimeMicroseconds':235914002,'UserTimeMicroseconds':1853801,'SystemTimeMicroseconds':869529,'SoftPageFaults':1237,'OSCPUWaitMicroseconds':1248845,'OSCPUVirtualTimeMicroseconds':2722131,'OSWriteBytes':4096,'OSReadChars':5817344,'OSWriteChars':1773568,'S3ReadMicroseconds':94480,'S3ReadRequestsCount':2,'S3ReadRequestsErrors':1,'S3WriteMicroseconds':77458343,'S3WriteRequestsCount':1000,'S3PutObject':1000,'WriteBufferFromS3Microseconds':78129835,'WriteBufferFromS3Bytes':332000,'WriteBufferFromS3WaitInflightLimitMicroseconds':7,'LogInfo':6}
```

If I try the same thing exporting to CSV `peak_threads_usage` goes to 1k and memory usage is also higher:

```
SELECT
    query,
    formatReadableSize(read_bytes),
    formatReadableSize(written_bytes),
    formatReadableSize(memory_usage) AS used_memory,
    peak_threads_usage,
    ProfileEvents['S3PutObject'] AS files,
    Settings,
    ProfileEvents
FROM system.query_log
WHERE (event_time >= (now() - toIntervalHour(1))) AND (query_id = 'a85c51db-7ed2-44bd-9b23-aebc3041f02b') AND (type > 1)
FORMAT Vertical

Query id: 827cd13b-3834-4f0c-95cb-9f6c1d4e97b9

Row 1:
──────
query:                             INSERT INTO FUNCTION s3('s3://test-sinks/test-threads/{_partition_id}.csv', 'REDACTED', '[HIDDEN]') PARTITION BY number SELECT * FROM numbers(1000)
formatReadableSize(read_bytes):    7.81 KiB
formatReadableSize(written_bytes): 7.81 KiB
used_memory:                       3.93 GiB
peak_threads_usage:                1003
files:                             1000
Settings:                          {'max_threads':'16','use_uncompressed_cache':'0','distributed_foreground_insert':'1','move_all_conditions_to_prewhere':'0','enable_multiple_prewhere_read_steps':'0','move_primary_key_columns_to_end_of_prewhere':'0','compile_expressions':'0','compile_aggregate_expressions':'1','enable_memory_bound_merging_of_aggregation_results':'0','cluster_for_parallel_replicas':'gcp_wadus_4','insert_deduplicate':'1','http_wait_end_of_query':'1','http_write_exception_in_output_format':'0','http_response_buffer_size':'104857600','joined_subquery_requires_alias':'0','max_bytes_before_external_group_by':'1442450940','max_execution_time':'0','max_expanded_ast_elements':'50000','join_algorithm':'hash','max_memory_usage':'5000000000','memory_usage_overcommit_max_wait_microseconds':'50000','log_query_threads':'0','formatdatetime_f_prints_single_zero':'1','formatdatetime_parsedatetime_m_is_month_name':'0','max_partitions_per_insert_block':'100','enable_lightweight_delete':'0','optimize_monotonous_functions_in_order_by':'0','enable_global_with_statement':'0','optimize_rewrite_sum_if_to_count_if':'0','distributed_ddl_entry_format_version':'3','enable_filesystem_cache':'1','enable_filesystem_cache_on_write_operations':'1','enable_filesystem_cache_log':'1','allow_prefetched_read_pool_for_remote_filesystem':'1','filesystem_prefetch_max_memory_usage':'5368709120','insert_keeper_max_retries':'15','insert_keeper_retry_initial_backoff_ms':'100','insert_keeper_retry_max_backoff_ms':'2000','input_format_null_as_default':'0'}
ProfileEvents:                     {'Query':1,'InsertQuery':1,'QueriesWithSubqueries':2,'SelectQueriesWithSubqueries':2,'IOBufferAllocs':6000,'IOBufferAllocBytes':6291834000,'ArenaAllocChunks':1,'ArenaAllocBytes':4096,'FunctionExecute':2,'TableFunctionExecute':2,'NetworkReceiveElapsedMicroseconds':4257,'NetworkSendElapsedMicroseconds':271711,'NetworkSendBytes':1496039,'InsertedRows':1000,'InsertedBytes':8000,'SelectedRows':1000,'SelectedBytes':8000,'ContextLock':1051,'ContextLockWaitMicroseconds':3,'RWLockAcquiredReadLocks':4,'RealTimeMicroseconds':35163037468,'UserTimeMicroseconds':1993555,'SystemTimeMicroseconds':1047662,'SoftPageFaults':17664,'OSCPUWaitMicroseconds':1916313,'OSCPUVirtualTimeMicroseconds':3000372,'OSReadChars':5585920,'OSWriteChars':1239040,'S3ReadMicroseconds':107173,'S3ReadRequestsCount':2,'S3ReadRequestsErrors':1,'S3WriteMicroseconds':66349635,'S3WriteRequestsCount':1000,'S3PutObject':1000,'WriteBufferFromS3Microseconds':67068731,'WriteBufferFromS3Bytes':3890,'WriteBufferFromS3WaitInflightLimitMicroseconds':3,'LogInfo':6}
```

**Comment by arthurpassos** [comment]
Continuing the [discussion](https://github.com/ClickHouse/ClickHouse/pull/76802#issuecomment-2722386456) here.

Copy from that thread:

I had a look at the issue you shared https://github.com/ClickHouse/ClickHouse/issues/51533.

Here are my findings so far: 

[PartitionedSink](https://github.com/ClickHouse/ClickHouse/blob/5313dc785a393066a66e2c6d30d575b0749aefed/src/Storages/PartitionedSink.cpp#L105) will create one Sink for each partition. The sink can be of three types: `StorageFileSink`, `StorageObjectStorageSink` or `StorageURLSink`. Except for StorageURLSink, the others will [create](https://github.com/ClickHouse/ClickHouse/blob/278c6da6177474a205284bc41efc7a19e4a80822/src/Storages/ObjectStorage/StorageObjectStorageSink.cpp#L74) a `ParallelFormattingOutputFormat` if `output_format_parallel_formatting` is set, which is  by default.

`ParallelFormattingOutputFormat`  will create two things: 

1. A thread pool that respects `max_threads` and allocates from the global thread pool. This pool will be used to spawn threads to process chunks in parallel.
2. A thread, not a pool, from the global thread pool to collect chunks that have been processed.

Both of these are problematic if many `Sink` objects are created. 

To address that, a couple of options come to mind:

1. `PartitionedSink` creates a shared thread pool and passes that down to its sinks AND `getOutputFormatParallelIfPossible`. This pool will be used for both the collector thread and the workr threads. **Not so simple to implement**
2. Somehow limit the amount of sinks that are created in [PartitionedSink](https://github.com/ClickHouse/ClickHouse/blob/5313dc785a393066a66e2c6d30d575b0749aefed/src/Storages/PartitionedSink.cpp#L105). This would attack the problem of bad partitioning leading to way too many threads. The downside is that  `PartitionedSink` has no notion of parallelism, it is agnostic to that. Code won't look pretty, we'll have to check for `output_format_parallel_formatting` in two places and etc.
3. Create a static thread pool for the ParallelFormattingOutputFormat::collector_thread and processor threads and use a server setting (can't be changed throughout the execution) to set the max threads. It might lead to starvation (e.g, a single bad query taking all the threads from the pool and blocking other good queries).
4. What you suggested: somehow limit the amount of `WriteBufferFromS3` that can be created at the same time. Just like the above, it might lead to starvation.

What do you think? Do you have a better suggestion?

**Comment by arthurpassos** [comment]
I have implemented a draft version of what the first approach would look like, but it is broken https://github.com/ClickHouse/ClickHouse/pull/77741.

**Comment by orloffv** [comment]
Hi @Algunenano ,

I’ve also encountered this problem in our production environment.

We’re running regular backups to S3 every 4 hours, and each backup increases the memory usage by ~1.5 GB on the replica performing the backup. Over ~3 days, the memory consumption grows until the host runs out of RAM and ClickHouse starts throwing memory limit exceeded errors on any queries.

Interestingly, the replicas where no backups are performed stabilize at around 18 GB of memory usage, while the backup-performing replica keeps growing indefinitely.

Do you have any ideas on what can be done right now to mitigate this? Are there any workarounds or recommended settings we could use to avoid memory leaks from backups?

Thanks in advance!

**Comment by orloffv** [comment]
We have two hosts in the same shard: one runs backups to S3 every 4 hours, and the other does not. The hosts are identical in data and load.

In the attached graph, the green line shows memory usage on the host with backups, and the yellow line shows memory on the host without backups. You can clearly see that every 4 hours, when the backup runs, there is a step increase in memory usage on the host with backups. Over time, this causes memory to grow until the host runs out of RAM and queries start failing with memory limit exceeded. Meanwhile, the host without backups has stable memory usage.

<img width="1473" height="373" alt="Image" src="https://github.com/user-attachments/assets/80f5b273-336c-4405-8e3f-776e4587acda" />

**Comment by azat** [comment]
>Kind of broken because it can lead to starvation

@arthurpassos I like the first option, but the starvation can indeed happen, we can disable parallel formatting in case of there are not enough threads (based on `max_threads`) to create at least 2 formatters and 1 collector for each partition, as for collector starvation we can create then from the global pool

**Comment by azat** [comment]
>In the attached graph, the green line shows memory usage on the host with backups, and the yellow line shows memory on the host without backups

@orloffv this issue is likely unrelated so it is better to create new one (if do not miss anything here)

Also does the `MemoryTracking` metric grows? Or RSS of the process as well? What about `jemalloc.allocated` metric? Also which version do you have?

If only `MemoryTracking` grows, then https://github.com/ClickHouse/ClickHouse/pull/83607 may indeed help

If not, can you please create another issue and report additional info there.

Thanks.

**Comment by orloffv** [comment]
i will create a ticket after tests with https://github.com/ClickHouse/ClickHouse/pull/83607

**Comment by telperions** [comment]
Faced similar memory issue that @Beetelbrox described. But not sure is it a full duplicate because I didn't notice any changes  in `OSThreadsTotal`. Maybe is is because I've tried it on 25.8 version.
Also in `query_log` I see 20 in `peak_threads_usage`, at the same time `length(thread_ids)` is big - 912. 
Any workaround ideas appreciated.
How to reproduce:

PARTITION BY 1 key
```
INSERT INTO FUNCTION s3(
        'https://my-bucket.s3.amazonaws.com/test_numbers_1/{_partition_id}/numbers.parquet',
        format=Parquet,
        partition_strategy='wildcard')
    PARTITION BY floor(number / 100000000)
SELECT
    randomPrintableASCII(randUniform(50, 500)) AS s1,
    number
FROM numbers(100000000)
    SETTINGS s3_create_new_file_on_insert = 1;

0 rows in set. Elapsed: 47.804 sec. Processed 100.00 million rows, 800.00 MB (2.09 million rows/s., 16.74 MB/s.)
Peak memory usage: 3.67 GiB.
```
PARTITION BY 10 keys
```
INSERT INTO FUNCTION s3(
        'https://my-bucket.s3.amazonaws.com/test_numbers_2/{_partition_id}/numbers.parquet',
        format=Parquet,
        partition_strategy='wildcard')
    PARTITION BY floor(number / 10000000)
SELECT
    randomPrintableASCII(randUniform(50, 500)) AS s1,
    number
FROM numbers(100000000)
    SETTINGS s3_create_new_file_on_insert = 1;

0 rows in set. Elapsed: 58.508 sec. Processed 100.00 million rows, 800.00 MB (1.71 million rows/s., 13.67 MB/s.)
Peak memory usage: 10.52 GiB.
```
PARTITION BY 100 keys
```
INSERT INTO FUNCTION s3(
        'https://my-bucket.s3.amazonaws.com/test_numbers_3/{_partition_id}/numbers.parquet',
        format=Parquet,
        partition_strategy='wildcard')
    PARTITION BY floor(number / 1000000)
SELECT
    randomPrintableASCII(randUniform(50, 500)) AS s1,
    number
FROM numbers(100000000)
    SETTINGS s3_create_new_file_on_insert = 1;

0 rows in set. Elapsed: 103.873 sec. Processed 100.00 million rows, 800.00 MB (962.71 thousand rows/s., 7.70 MB/s.)
Peak memory usage: 27.92 GiB.
```
